### DISCUSSION TITLE
hyperf2.x或swoole环境使用easywehat6.x

### DISCUSSION THREAD

**Discussion by awen1988** [w7corp/easywechat#2449] [General]
注意：以下操作经过本人在生产环境不完全验证，因为我本身仅仅是一个业务人员，所以涉及到专业的知识性错误欢迎大家补充指正！！！
环境：
1、如果使用的是hyperf2.x则PHP必须是8.0.x，因为hyperf2.x不支持8.1以上，easywechat6.x不支持8.0以下
2、swoole截止到我测试通过，最新的发布版本4.8.x是不支持的，原因是curl函数在swoole携程化下有bug，具体的报错可以自行编译测试，总之就是与easywechat6.x使用的symfony/http-client包中的某些curl不兼容，实际上hyperf提供了携程的Guzzle HTTP客户端但是easywechat6.x在调研之初就否定了，参考[https://github.com/w7corp/easywechat/issues/2328](https://github.com/w7corp/easywechat/issues/2328)，然后6.x实际上也只封装了核心部分功能，具体业务接口还是需要自己手动写请求，所以如果你hyperf非常牛，guzzle用得也非常6完全可以把6.x的功能抄过来直接用就好了，这个不在本文的讨论范围。
但是swoole的master分支编译后可以无缝支持，原因是这些bug已经修复了，可以看一下最新的提交记录

已经测试通过的功能：
公众号、开放平台、支付，小程序、企业微信因为业务暂时没有用到暂未测试

原理：跟5.x在hyperf下使用类似，因为实际上这样的API相互调用的sdk就是发请求处理响应，而优秀的easywechat6.x提供了完全自定义的能力，包括核心的http请求客户端，所以你完全可以根据自己的实际需求来替换，当前全都换了她就不是那个她了，哈哈～
客户端：es6.x要求是实现了Symfony\Contracts\HttpClient\HttpClientInterface接口的，所以如果你无法升级swoole的版本，可以通过替换客户端的方式实现，写一个实现这个接口的类，奈何我不会，幸好swoole5.x已经兼容了所以本步骤跳过
请求：hyperf的请求类跟es6.x要求的请求类是实现了同一套接口的，可以直接替换使用且必须替换，否则es6.x拿不到请求参数
缓存：hyperf的缓存同样跟es6.x要求的缓存类是实现同一套接口的，可直接使用，建议替换，默认的文件缓存根据5.x的经验隔一段时间或者并发较大的情况下或造成缓存错误的问题。

开放平台：
```php
// $platformModel 是开放平台模型类，简单来说是从数据库中取出来的，如果用配置文件也可以
public function initApplication(PlatformModel $platformModel)
    {
        // 实例化应用
        $application = make(Application::class, [[
            'app_id' => $platformModel->app_id,
            'secret' => $platformModel->app_secret,
            'token' => $platformModel->app_token,
            'aes_key' => $platformModel->app_key
        ]]);

        // region 替换请求[必须替换] +++++
        $request = ApplicationContext::getContainer()->get(RequestInterface::class);
        $application->setRequest($request);
        // endregion

        // region 替换缓存[可选，建议替换] +++++
        $cache = ApplicationContext::getContainer()->get(CacheInterface::class);
        $application->setCache($cache);
        // endregion

        return $application;
    }
```

开放平台代公众号实现业务，这步操作必须要有，原因是开放平台也需要去兑换公众号的应用实例，所以兑换出来的实例也需要替换响应的请求，以下可供参考，需要根据自己的业务修改。
```php
public function getOfficialApplication(AuthorizeModel $authorizeModel): \EasyWeChat\OfficialAccount\Application
    {
        $platformModel = $authorizeModel->platform;
        $platformApplication = $this->initApplication($platformModel);
        $cache = ApplicationContext::getContainer()->get(CacheInterface::class);
        $cacheName = md5($platformModel->id.'_'.$authorizeModel->app_id.'_authorizer_access_token');
        $authorizerAccessToken = $cache->get($cacheName, '');
        if (!$authorizerAccessToken) {
            $refreshAuthorizerToken = $platformApplication->refreshAuthorizerToken($authorizeModel->app_id, $authorizeModel->refresh_token);
            $authorizerAccessToken = $refreshAuthorizerToken['authorizer_access_token'];
            // 写入缓存
            $cache->set($cacheName, $authorizerAccessToken, $refreshAuthorizerToken['expires_in']);
        }
        $officialAccessToken = make(AuthorizerAccessToken::class, [$authorizeModel->app_id, $authorizerAccessToken]);
        // 公众号应用
        $officialApplication = $platformApplication->getOfficialAccount($officialAccessToken);
        // region 替换请求 +++++
        $request = ApplicationContext::getContainer()->get(RequestInterface::class);
        $officialApplication->setRequest($request);
        // endregion
        // region 替换缓存 +++++
        $officialApplication->setCache($cache);
        // endregion
        // 设置客户端
        $officialApplication->setClient($platformApplication->getClient());
        // var_dump($officialApplication->getAccessToken());
        // $client = $officialApplication->createClient();
        // $officialApplication->setClient($client);
        return $officialApplication;
    }
```

公众号：这是直接使用公众号后台配置接口的情况，为了兼容，写了开放平台的东西，如果你完全没用开放平台可以直接替换请求合缓存就能用。
```php
public function initApplication(OfficialModel $officialModel, PlatformModel $platformModel = null)
    {
        $authorizeModel = null;
        // 如指定开放平台或公众号自身缺少必要属性则尝试开放平台实例
        if ($platformModel || !$officialModel->app_secret || !$officialModel->app_token) {
            if ($platformModel) {
                $authorizeModel = $this->authorizeService->model->where([
                    ['platform_id', '=', $platformModel->id],
                    ['app_id', '=', $officialModel->app_id]
                ])->firstOrFail();
            } else {
                $authorizeModel = $this->authorizeService->model->where([
                    ['app_id', '=', $officialModel->app_id]
                ])->firstOrFail();
            }
        }
        if ($authorizeModel instanceof AuthorizeModel) {
            return $this->platformService->getOfficialApplication($authorizeModel);
        }
        // 实例化应用
        $application = make(Application::class, [[
            'app_id' => $officialModel->app_id,
            'secret' => $officialModel->app_secret,
            'token' => $officialModel->app_token,
            'aes_key' => $officialModel->app_key,
        ]]);

        // region 替换请求[必须替换] +++++
        $request = ApplicationContext::getContainer()->get(RequestInterface::class);
        $application->setRequest($request);
        // endregion

        // region 替换缓存[可选，建议替换] +++++
        $cache = ApplicationContext::getContainer()->get(CacheInterface::class);
        $application->setCache($cache);
        // endregion

        return $application;
    }
```
支付：支付仅测试通过v2，暂未测试v3，如有问题我会补充
```php
public function initApplication(MerchantModel $merchantModel)
    {
        $application = make(Application::class, [[
            // 商户号
            'mch_id' => $merchantModel->mch_id,
            // 商户证书
            'certificate' => BASE_PATH.'/cert/'.$merchantModel->mch_id.'/apiclient_cert.pem',
            'private_key' => BASE_PATH.'/cert/'.$merchantModel->mch_id.'/apiclient_key.pem',
            // v2 API 密钥
            'v2_secret_key' => $merchantModel->pay_key,
            // v3 API 密钥
            'secret_key' => $merchantModel->v3_pay_key,

        ]]);

        // region 替换请求 +++++
        $request = ApplicationContext::getContainer()->get(RequestInterface::class);
        $server = make(Server::class, [
            $application->getMerchant(),
            $request
        ]);
        $application->setServer($server);
        // endregion

        return $application;
    }
```

**Comment by lshaofan** [comment]
你好 ，按照你的方法 我这里直接走到了这个错误，没搞明白
[ERROR] Failed to decode request contents.[62] in /www/wwwroot/server/vendor/w7corp/easywechat/src/Kernel/Message.php

**Reply by awen1988** [reply]
这是哪一步呢

**Reply by lshaofan** [reply]
处理服务端这里
$server = $app->getServer();
return $server->serve();
实例化开放平台跟你的一样步骤，
感觉就是请求对象替换的问题，按照
[wechat示例](https://easywechat.com/6.x/open-platform/examples.html)
改成SymfonyRequest也是同样的报错信息，
环境
PHP 8.0.14 (cli) (built: Dec 18 2021 21:57:50) ( NTS )
Copyright (c) The PHP Group
Zend Engine v4.0.14, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.13, Copyright (c), by Zend Technologies


swoole

Swoole => enabled
Author => Swoole Team <team@swoole.com>
Version => 4.8.5
Built => Dec 26 2021 01:13:34
coroutine => enabled with boost asm context
epoll => enabled
eventfd => enabled
signalfd => enabled
spinlock => enabled
rwlock => enabled
openssl => OpenSSL 1.1.1l  24 Aug 2021
dtls => enabled
http2 => enabled
json => enabled
curl-native => enabled
pcre => enabled
zlib => 1.2.11
brotli => E16777225/D16777225
mutex_timedlock => enabled
pthread_barrier => enabled
async_redis => enabled

Directive => Local Value => Master Value
swoole.enable_coroutine => On => On
swoole.enable_library => On => On
swoole.enable_preemptive_scheduler => Off => Off
swoole.display_errors => On => On
swoole.use_shortname => Off => Off
swoole.unixsock_buffer_size => 8388608 => 8388608

**Reply by awen1988** [reply]
我测试过4.8.x有问题，我是编译的5.0-dev版本

**Reply by jiangslee** [reply]
https://github.com/limingxinleo/easywechat-classmap

![image](https://user-images.githubusercontent.com/27892476/170814964-3e0f0a94-02d5-4433-bc28-ea5b8370438c.png)

**Reply by lshaofan** [reply]
https://github.com/limingxinleo/easywechat-classmap
这个require   还是这个报错

**Reply by jiangslee** [reply]
> https://github.com/limingxinleo/easywechat-classmap
> 这个require 还是这个报错

你的报错是直接用浏览器请求触发的吗？？

**Reply by Mr12312** [reply]
Hi jiangslee,我用这包时候发现“use EasyWeChat\Kernel\HttpClient\RequestUtil;”不存在easywechat6.x，这个做了什么处理的？谢谢。
![捕获12](https://user-images.githubusercontent.com/29160167/176353034-6ba0d328-faa1-4272-8e4d-619f1cbffd68.PNG)

**Reply by jiangslee** [reply]
建议你去掉composer代理，然后重新安装一下包。
```
composer config -g --unset repos.packagist
```
> 有些composer代理镜像不全(如阿里云)，使用以上命令可删除composer代理

**Comment by a102874** [comment]
V3支付，使用提到处理，请求400 Bad Request returned
